test/integration/isolated/suite_test.go

```diff
@@ -0,0 +1,346 @@
+
+	var l sync.RWMutex
+	tenv.BeforeEachFeature(
+		// TODO: Prevent a data race by using a mutex explicitly when first creating the client.
+		// Related: https://github.com/Kong/kubernetes-ingress-controller/issues/4848
+		func(ctx context.Context, c *envconf.Config, t *testing.T, f features.Feature) (context.Context, error) {
+			l.Lock()
+			defer l.Unlock()
+			_, err = c.NewClient()
+			return ctx, err
+		},
+	)
+
+	code := tenv.Run(m)
+	defer func() {
+		os.Exit(code)
+	}()
+
+	if testenv.IsCI() {
+		fmt.Printf("INFO: running in ephemeral CI environment, skipping cluster %s teardown\n", env.Cluster().Name())
+	} else {
+		ctx, cancel := context.WithTimeout(context.Background(), test.EnvironmentCleanupTimeout)
+		defer cancel()
+		helpers.ExitOnErr(ctx, helpers.RemoveCluster(ctx, env.Cluster()))
+	}
+}
+
+func featureSetup(opts ...helpers.ControllerManagerOpt) func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+	return func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+		// TODO: this is temporary to allow things like:
+		// clusters.ApplyManifestByYAML(ctx, cluster, s)
+		// in tests.
+		// Ideally this wouldn't be possible because it allows tests to break
+		// a shared cluster but we don't have means to use kubectl against a cluster
+		// without including a 3rd party package.
+		// We should remove the cluster from the context as the last setup step.
+		cluster := GetClusterFromCtx(ctx)
+
+		runID := GetRunIDFromCtx(ctx)
+
+		ctx, err := createNSForTest(ctx, cfg, t, runID)
+		if !assert.NoError(t, err) {
+			return ctx
+		}
+
+		namespace := GetNamespaceForT(ctx, t)
+
+		// Logger needs to be configured before anything else happens.
+		// This is because the controller manager has a timeout for
+		// logger initialization, and if the logger isn't configured
+		// after 30s from the start of controller manager package init function,
+		// the controller manager will set up a no op logger and continue.
+		// The logger cannot be configured after that point.
+		logger, logOutput, err := testutils.SetupLoggers("trace", "text")
+		if !assert.NoError(t, err, "failed to setup loggers") {
+			return ctx
+		}
+		if logOutput != "" {
+			t.Logf("writing manager logs to %s", logOutput)
+		}
+
+		t.Logf("setting up test environment")
+		kongbuilder, extraControllerArgs, err := helpers.GenerateKongBuilder(ctx)
+		if !assert.NoError(t, err) {
+			return ctx
+		}
+		if testenv.KongImage() != "" && testenv.KongTag() != "" {
+			fmt.Printf("INFO: custom kong image specified via env: %s:%s\n", testenv.KongImage(), testenv.KongTag())
+		}
+
+		// Pin the Helm chart version.
+		kongbuilder.WithHelmChartVersion(testenv.KongHelmChartVersion())
+		kongbuilder.WithNamespace(namespace)
+		kongbuilder.WithName(NameFromT(t))
+		kongbuilder.WithAdditionalValue("readinessProbe.initialDelaySeconds", "1")
+
+		kongAddon := kongbuilder.Build()
```

Comment:
What is the preferred way to modify the configuration of the Kong add-on per test e.g. to do something like that
```go
WithProxyEnvVar("PROXY_LISTEN", `0.0.0.0:8000 http2\, 0.0.0.0:8443 http2 ssl`).
```
I have already a use-case for such a thing in 
- https://github.com/Kong/kubernetes-ingress-controller/issues/4273

Reply:
Probably this is the  way to go now. I haven't looked into it.

Reply:
So maybe `helpers.ControllerManagerOpt` should be more generic to allow adding an option func to set other stuff like env values in Kong builder. Because now it is limited to set manager options only. Function `featureSetup` is used in every isolated test so it should allow detailed customization. 

It can be done later, but it will require changing every occurrence of `helpers.ControllerManagerOpt`

Reply:
So maybe `helpers.ControllerManagerOpt` should be more generic to allow adding an option func to set other stuff like env values in Kong builder. Because now it is limited to set manager options only. Function `featureSetup` is used in every isolated test so it should allow detailed customization. 

It can be done later, but it will require changing every occurrence of `helpers.ControllerManagerOpt`

Reply:
👍 Good idea, feel free to submit an issue which would capture this. For now there are just a handful of usages of `helpers.ControllerManagerOpt` so I believe we're good but making this more generic definitely makes sense. (as long as we have use case for that genericness 🙃 )

Reply:
🚀